src/string/is_valid_email_address.rs

```diff
@@ -0,0 +1,255 @@
+const MAX_LOCAL_PART_LENGTH: u32 = 64;
+const MAX_DOMAIN_LENGTH: u32 = 255;
+const CHARACTERS: &str = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz1234567890";
+const SYMBOLS: &str = "!#$%&'*+-/=?^_`{|}~";
+const HEX_CHARACTERS: &str = "0123456789abcdef";
+
+fn clear_quotes(input: &str) -> String {
+    let mut quote_count = input.starts_with('"') as u32;
+    let mut new_local_part: String = String::from(input.chars().next().unwrap());
+    for i in 1..input.len() {
+        if input.chars().nth(i).unwrap() == '"' && input.chars().nth(i - 1).unwrap() != '\\' {
+            quote_count += 1;
+
+            if !new_local_part.starts_with('"') && quote_count != 1 {
+                new_local_part.push('"');
+            }
+        }
+
+        if quote_count % 2 == 0 {
+            new_local_part.push(input.chars().nth(i).unwrap());
+        }
+    }
+
+    new_local_part
+}
+
+fn has_bad_dots(part: &str) -> bool {
+    part.starts_with('.') || part.ends_with('.') || part.contains("..")
+}
+
+fn contains_legal_local_characters(local_part: &str) -> bool {
+    for item in local_part.chars() {
+        if !CHARACTERS.contains(item) && !SYMBOLS.contains(item) && item != '.' && item != '"' {
+            return false;
+        }
+    }
+    true
+}
+
+fn contains_legal_domain_characters(domain: &str) -> bool {
+    for item in domain.chars() {
+        if !CHARACTERS.contains(item) && item != '-' && item != '.' && !"[]:".contains(item) {
+            return false;
+        }
+    }
+    true
+}
+
+fn has_valid_quotes(part: &str) -> bool {
+    for i in 1..part.len() {
+        if part.chars().nth(i - 1).unwrap() == '"' && part.chars().nth(i).unwrap() == '"' {
+            let proceeding_quote = part.chars().nth(i + 1);
+            if proceeding_quote.is_some() && proceeding_quote.unwrap() != '.' {
+                return false;
+            }
+        }
+    }
+    true
+}
+
+fn is_valid_ipv4(ip: &str) -> bool {
+    let parts = ip
+        .split('.')
+        .map(|x| x.to_string())
+        .collect::<Vec<String>>();
+    if parts.len() != 4 {
+        return false;
+    }
+
+    for value in parts {
+        let value = value.parse::<u32>();
+        if value.is_err() || value.unwrap() > 255 {
+            return false;
+        }
+    }
+
+    true
+}
+
+fn is_valid_ipv6(ip: &str) -> bool {
+    let parts = ip
+        .split(':')
+        .map(|x| x.to_string())
+        .collect::<Vec<String>>();
+
+    // we check for the length 9 here as the IPv6 counts as one of the parts
+    if parts.len() != 9 || parts.first().unwrap() != "IPv6" {
+        return false;
+    }
+
+    for value in &parts[1..parts.len()] {
+        if value.is_empty() || value.len() > 4 {
+            return false;
+        }
+
+        for chr in value.chars() {
+            if !HEX_CHARACTERS.contains(chr.to_ascii_lowercase()) {
+                return false;
+            }
+        }
+    }
+
+    true
+}
```

Comment:
What do you think about testing these functions directly?

Reply:
I renamed `is_valid_ipv4` and `is_valid_ipv6` to `is_ipv4` and `is_ipv6` respectively, but for the rest of the functions which contained the word 'valid', I moved the valid to the middle of the function name to make the code more consistent while staying readable.

In addition I added tests for `is_ipv4` and `is_ipv6`, note that some ips were taken from the `is_valid_email_address` tests but I still kept those tests in as I thought they may still be good tests to have, they can be removed if you want. Also, due to adding more tests, I made two new macros, but the way I did it may not be the most efficient way of doing so, but it shouldn't matter too much as it is just for the tests.